### PR TITLE
Feature 10

### LINKED ISSUE #10: As a live cell I will die if I have fewer than two live neighbours

### PR DESCRIPTION
**Pull request by fjank**
resolves #10

### CODE REVIEW

**Review by netbrain — DISMISSED** [review]
Nothing wrong with the code, other than it feels a bit over-engineered for the first issue. I would have preferred a simpler implementation, the least complex solution to make the test pass, and evolve the code as new requirements are added to the codebase. But that being said, it is quite typical for developers to get ahead of themselves :) me included!

Basically letting the tests drive the architecture of the code with a focus of only solving the given task at hand at any point in time.

As for cucumber, it's basically a way of giving business people a way to define tests and letting developers implement them. If i'm to say anything about cucumber it would be that I'd prefer a way to test grid's more easily. As you probably want it to be both writable and readable for the business people. An example would be:

```
Given the grid:
  *..
  ...
  ...

Then the next generation should look like:
  ...
  ...
  ...                        

```

However there might be technical limitations to cucumber which doesn't make this possible, idk.

**Comment by fjank** [comment]
Agree to both comments, will restructure the cucumber tests, and massage the code to make it the least complex and only keep code to make the tests green. Pure BDD/TDD driven code!

**Comment by netbrain** [comment]
As an example. Say you were working on #4 instead. And say you write the tests first, then given the test (pseudocode):

```
testShouldGiveTextualRepresentationOfGrid(){
 Grid g = new Grid(3,3);
 assertEquals(
    "...\n"+
    "...\n"+
    "...\n",
    g.toString()
  );
}
```

Now for the kicker, the following is a perfectly fine implementation for this first test:

```
class Grid {
  public Grid(int w, int h){}
  public String toString(){
    return "...\n...\n...\n";
  }
}
```

So as you add business logic you should try to solve the tests as simple as possible and force yourself to create tests that ultimately enforces correct behavior.

So for the next test you would probably try to create a grid with a different size, say 4 by 4. Which in turn would have to actually store that information and depending on that output the correct string for 3x3 and 4x4. after that you would probably add notation for living cells. 

It's like you'r trying to play two roles, the tester and the implementor, where the implementor is constantly trying to trick the tester and create silly logic that actually wouldn't hold in real life. And then the tester needs to counteract those implementations with more and more elaborate tests.

Now ofc. in real life you probably don't do this even as a seasoned TDD practioner, however I still think it's useful to do this in honing one's skills. It forces you to think deeply on the current situation.

**Comment by fjank** [comment]
Gah! I went to far again, I see where you want to go. I will add another commit that I think will represent this even better, removing most of the scenarios (leaving only one), and the absolutely simplest code to make it work.

File: src/test/resources/features/two-neighbour-rule.feature [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+  Scenario: 3x3 grid with three neighbours
+    All cells has 2 or more neighbours, all will live on.
+    Given the grid:
+    """
+      ***
+      .*.
+      ...
+    """
+    When the generation is incremented
```
Comment (netbrain) [inline src/test/resources/features/two-neighbour-rule.feature]:
```suggestion
```
Can leave it out. The "Then" should handle generation incrementation when you get to that

**Review by netbrain — CHANGES_REQUESTED** [review]

File: src/test/java/features/TwoNeighbourRuleSteps.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+     * Creates a new StepDefinition.
+     */
+    public TwoNeighbourRuleSteps() {
+        Given("the grid:", (String gridDef) -> {
+            grid = new Grid();
+        });
+
+        When("the generation is incremented", () -> {
+
+        });
```
Comment (fjank) [inline src/test/java/features/TwoNeighbourRuleSteps.java]:
```suggestion

```

**Review by fjank — COMMENTED** [review]
should also remove the implementation for the When step.

**Review by netbrain — APPROVED** [review]

**Pull request merged**